pkg/cloudsql/queries.go

```diff
@@ -0,0 +1,140 @@
+}
+
+type MemoryMetrics struct {
+	Total               int64
+	Used                int64
+	Freeable            int64
+	UsedPercentage      float64
+	AvailablePercentage float64
+}
+
+func GetMemoryMetrics(cloudMonitoringClient *CloudMonitoringClient, projectID string, databaseName string) (MemoryMetrics, error) {
+	totalMemory, err := GetMemoryTotal(cloudMonitoringClient, projectID, databaseName)
+	if err != nil {
+		return MemoryMetrics{}, err
+	}
+	usedPercentage, err := getMemoryComponents(cloudMonitoringClient, projectID, databaseName, "Usage")
+	if err != nil {
+		return MemoryMetrics{}, err
+	}
+	cachePercentage, err := getMemoryComponents(cloudMonitoringClient, projectID, databaseName, "Cache")
+	if err != nil {
+		return MemoryMetrics{}, err
+	}
+	freePercentage, err := getMemoryComponents(cloudMonitoringClient, projectID, databaseName, "Free")
+	if err != nil {
+		return MemoryMetrics{}, err
+	}
+
+	return MemoryMetrics{
+		Total:               totalMemory.GetInt64Value(),
+		Used:                int64(float64(totalMemory.GetInt64Value()) * usedPercentage.GetDoubleValue() / 100.),
+		UsedPercentage:      usedPercentage.GetDoubleValue(),
+		Freeable:            int64(float64(totalMemory.GetInt64Value()) * (freePercentage.GetDoubleValue() + cachePercentage.GetDoubleValue()) / 100.),
+		AvailablePercentage: freePercentage.GetDoubleValue() + cachePercentage.GetDoubleValue(),
+	}, nil
+}
+
+func getLatestMetricValue(cloudMonitoringClient *CloudMonitoringClient, projectID string, metricName string, resourceLabels []QueryLabel, metricLabels []QueryLabel) (monitoringpb.TypedValue, error) {
+	endTime := time.Now().UTC()
+	// get the last _complete_ minute of data
+	startTime := endTime.Add(-10 * time.Minute)
+
+	filter := fmt.Sprintf(`metric.type = "%s"`, metricName)
+
+	for _, label := range resourceLabels {
+		filter = fmt.Sprintf(`%s AND resource.label.%s = "%s"`, filter, label.Name, label.Value)
+	}
+
+	for _, label := range metricLabels {
+		filter = fmt.Sprintf(`%s AND metric.labels.%s = "%s"`, filter, label.Name, label.Value)
+	}
+
+	req := &monitoringpb.ListTimeSeriesRequest{
+		Name:   fmt.Sprintf("projects/%s", projectID),
+		Filter: filter,
+		Interval: &monitoringpb.TimeInterval{
+			StartTime: &timestamppb.Timestamp{Seconds: startTime.Unix()},
+			EndTime:   &timestamppb.Timestamp{Seconds: endTime.Unix()},
+		},
+		View: monitoringpb.ListTimeSeriesRequest_FULL,
+	}
+
+	// TODO: better contexts
+	it := cloudMonitoringClient.client.ListTimeSeries(cloudMonitoringClient.ctx, req)
+
+	// we expect there to be exactly one result
+	resp, err := it.Next()
+	if err == iterator.Done {
+		return monitoringpb.TypedValue{}, fmt.Errorf("Oh no, we didn't get any results!")
+	}
+	// TODO: what other errors can this actually return?
+	if err != nil {
+		log.Fatalf("Could not read time series value: %v", err)
+	}
+
+	return *resp.GetPoints()[0].GetValue(), nil
```

Comment:
`GetValue()` returns a pointer? I've been burned by not null checking it as it was empty once -_-

GG glorious GO